### PR TITLE
Prerendering

### PR DESCRIPTION
**Pull request by k88hudson**
This patch adds prerendering in a few commits:

- a5356f7: This is just some basic refactoring to not crash in node (using `global` for globals, etc.)
- 2193572: This changes the state rehydration to work with a new action, `NEW_TAB_STATE_REQUEST`, instead of `NEW_TAB_LOAD`. This is needed for prerendering in order to ensure the first render matches the prerendered state.
- 45357fb: This adds the `buildmc:html` task, and all the basic prerendering tasks.
- 686518a: This adds a pref to determine whether or not the pre-rendered version should be used. I need to actually add a patch to `aboutNewTabService` to make use of this.
- 209f43dd: This is a small change to `<Topics />` to allow it to render but be empty. This is to ensure things don't move around when the pre-rendered version switches to React.

### CODE REVIEW

File: system-addon/common/PerfService.jsm [hunk 1]
```diff
@@ -1,26 +1,26 @@
...
   usablePerfObj = Services.appShell.hiddenDOMWindow.performance;
-} else { // we must be running in content space
+} else if (typeof performance !== "undefined") {
+  // we must be running in content space
   usablePerfObj = performance;
+} else {
+  usablePerfObj = {
+    now() {},
+    mark() {}
+  };
```
Comment (dmose) [inline system-addon/common/PerfService.jsm]: I'm assuming this is needed so that it runs under node.  A comment indicating that would nice.

File: system-addon/content-src/lib/constants.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
 module.exports = {
   // constant to know if the page is about:newtab or about:home
-  IS_NEWTAB: document.documentURI === "about:newtab"
+  IS_NEWTAB: global.document && global.document.documentURI === "about:newtab"
```
Comment (dmose) [inline system-addon/content-src/lib/constants.js]: What is it that keeps this code running in the browser?  Webpack magic?  Do we assign global=Window somewhere?

File: system-addon/content-src/components/Base/Base.jsx [hunk 3]
```diff
@@ -22,6 +23,9 @@ function addLocaleDataForReactIntl({locale}) {
 
 class Base extends React.Component {
   componentDidMount() {
+    // Request state after the first render
+    this.props.dispatch(ac.SendToMain({type: at.NEW_TAB_STATE_REQUEST}));
+
```
Comment (dmose) [inline system-addon/content-src/components/Base/Base.jsx]: A comment explaining why we do this extra round trip (e.g. for pre-rendering), perhaps even in a bit of details, would likely be useful to future code readers.

File: system-addon/lib/NewTabInit.jsm [hunk 4]
```diff
@@ -15,7 +15,7 @@ this.NewTabInit = class NewTabInit {
   onAction(action) {
     let newAction;
     switch (action.type) {
-      case at.NEW_TAB_LOAD:
+      case at.NEW_TAB_STATE_REQUEST:
```
Comment (dmose) [inline system-addon/lib/NewTabInit.jsm]: This wants a unit test, I think.  Maybe a look at the code coverage numbers would show other changes that want one too?

File: system-addon/content-src/components/Base/Base.jsx (hunk 3 above)
Comment (dmose) [reply system-addon/content-src/components/Base/Base.jsx]: A unit test for this would be good too; https://github.com/airbnb/enzyme/blob/master/docs/api/mount.md has an example using sinon that should be similar.

File: bin/render-activity-stream-html.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    <script src="${options.baseUrl}vendor/react-redux.js"></script>
+    <script src="${options.baseUrl}data/content/activity-stream.bundle.js"></script>
+  </body>
+</html>
+`;
+}
+
+function templateJs(state) {
+  return `// Note - this is a generated file.
+  window.gActivityStreamPrerenderedState = ${JSON.stringify(state, null, 2)};
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: This wants a comment describing what it's doing and why.  Probably in the rendered file so that people who find that artifact quickly see what is.

File: bin/render-activity-stream-html.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+`;
+}
+
+const args = require("minimist")(process.argv.slice(2), {alias: {baseUrl: "b", title: "t", locale: "l", direction: "d"}});
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: This is a bit hard to skim.  Maybe break it up into two lines, and explain the magic constant of 2 in a comment?

File: bin/render-activity-stream-html.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+const options = Object.assign({}, DEFAULT_OPTIONS, args || {});
+const {html, state} = prerender(options.locale);
+
+fs.writeFileSync(HTML_FILE_PATH, templateHTML(options)); // eslint-disable-line no-sync
+fs.writeFileSync(PRERENDERED_HTML_FILE_PATH, templateHTML(options, html)); // eslint-disable-line no-sync
+fs.writeFileSync(INITIAL_STATE_JS_FILE_PATH, templateJs(state)); // eslint-disable-line no-sync
+
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: It'd improve readaibility to just put /* eslint-disable no-sync */ at the top of the file.

File: system-addon/content-src/components/Base/Base.jsx [hunk 8]
```diff
@@ -46,11 +54,18 @@ class Base extends React.Component {
...
+    if (!props.isPrerendered && !initialized) {
       return null;
     }
 
-    return (<IntlProvider key={locale} locale={locale} messages={strings}>
+    // Note: the key on IntlProvider must be static in order to not blow away
+    // all elements on a locale change (such as after preloading).
+    // See https://github.com/yahoo/react-intl/issues/695 for more info.
+    return (<IntlProvider key="STATIC" locale={locale} messages={strings}>
```
Comment (Mardak) [inline system-addon/content-src/components/Base/Base.jsx]: This change seems to cause the search placeholder and migration texts to not appear on the first about:home. (You can also test with `./mach run about:newtab`) Changing back to key={locale} avoids this:
image

My guess is that `LOCALE_UPDATED` happens at an unexpected time?

Comment (Mardak) [reply system-addon/content-src/components/Base/Base.jsx]: Looks like passing in a dummy `locale` prop, e.g., `<Search locale={locale} />`, makes it `render` again

File: system-addon/content-src/lib/constants.js (hunk 2 above)
Comment (k88hudson) [reply system-addon/content-src/lib/constants.js]: `global` is `window` in webpack, yes

**Comment by Mardak** [comment]
I pushed to talos moving the activity-stream-prerendered.html to activity-stream.html:
https://treeherder.mozilla.org/perf.html#/compare?originalProject=mozilla-central&originalRevision=c959327c6b75cd4930a6ea087583c38b805e7524&newProject=try&newRevision=099db27cebe866024186aac8f4f0274c60839d09&framework=1&filter=7%20opt

Nothing significant so far although tabpaint shows 6% improvement but low confidence after 10 before and 10 after runs.

File: system-addon/data/content/activity-stream.html [hunk 9]
```diff
@@ -1,24 +0,0 @@
...
-    <meta http-equiv="Content-Security-Policy-Report-Only" content="script-src 'unsafe-inline'; img-src http: https: data: blob:; style-src 'unsafe-inline'; child-src 'none'; object-src 'none'; report-uri https://tiles.services.mozilla.com/v4/links/activity-stream/csp">
-    <title></title>
-    <link rel="icon" type="image/png" id="favicon" href="chrome://branding/content/icon32.png"/>
-    <link rel="stylesheet" href="chrome://browser/content/contentSearchUI.css" />
-    <link rel="stylesheet" href="resource://activity-stream/data/content/activity-stream.css" />
-  </head>
-  <body class="activity-stream">
-    <div id="root"></div>
-    <div id="snippets-container">
-      <div id="snippets"></div>
```
Comment (Mardak) [inline system-addon/data/content/activity-stream.html]: We'll still need these `snippets` related div#ids, yeah?

**Comment by Mardak** [comment]
Still nothing significant (and maybe even a slight tart regression?)

I also pushed to try with aboutHome enabled:
https://treeherder.mozilla.org/perf.html#/compare?originalProject=mozilla-central&originalRevision=c959327c6b75cd4930a6ea087583c38b805e7524&newProject=try&newRevision=0c03e6c544e7a8a222b775b009110932d2a16cf9&framework=1

File: system-addon/lib/NewTabInit.jsm (hunk 4 above)
Comment (k88hudson) [reply system-addon/lib/NewTabInit.jsm]: Currently this file isn't being tested (which is an oversight), do you mind if I file a follow-up for this?

Comment (dmose) [reply system-addon/lib/NewTabInit.jsm]: Because of the high-value in getting this into the tree sooner rather than later, yes.

File: bin/render-activity-stream-html.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    <meta charset="utf-8">
+    <meta http-equiv="Content-Security-Policy-Report-Only" content="script-src 'unsafe-inline'; img-src http: https: data: blob:; style-src 'unsafe-inline'; child-src 'none'; object-src 'none'; report-uri https://tiles.services.mozilla.com/v4/links/activity-stream/csp">
+    <title>${options.title}</title>
+    <link rel="icon" type="image/png" id="favicon" href="chrome://branding/content/icon32.png"/>
+    <link rel="stylesheet" href="chrome://browser/content/contentSearchUI.css" />
+    <link rel="stylesheet" href="${options.baseUrl}data/content/activity-stream.css" />
+    <link rel="icon" type="image/svg+xml" href="${options.baseUrl}img/newtab-icon.svg">
+  </head>
+  <body>
+    <div id="root">${isPrerendered ? html : ""}</div>${isPrerendered ? `
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: Having so much of the semantics come from punctuation, and specifically nested template strings, makes this a bit hard to follow.  Pull out the <script> for stream-initial-state into a const higher up in the script?

File: bin/render-activity-stream-html.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
+const fs = require("fs");
+const path = require("path");
+const prerender = require("./prerender");
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: Add a comment here about what this is and how it's built?

File: bin/render-activity-stream-html.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    <script src="${options.baseUrl}vendor/react-intl.js"></script>
+    <script src="${options.baseUrl}vendor/redux.js"></script>
+    <script src="${options.baseUrl}vendor/react-redux.js"></script>
+    <script src="${options.baseUrl}data/content/activity-stream.bundle.js"></script>
+  </body>
+</html>
+`;
+}
+
+function templateJs(state) {
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: JSDoc here for first-time readers would be helpful, mostly for the info about "state"

File: system-addon/content-src/activity-stream-prerender.jsx [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  if (locale) {
+    if (!allStrings[locale]) {
+      throw new Error(`Tried to get strings for ${locale} but none were found.`);
+    }
+    strings = allStrings[locale];
+  } else {
+    Object.keys(allStrings["en-US"]).forEach(key => {
+      strings[key] = " ";
+    });
+  }
```
Comment (dmose) [inline system-addon/content-src/activity-stream-prerender.jsx]: @Mardak, does this need a spin-off bug to properly do individual string fallback correctly?

Comment (k88hudson) [reply system-addon/content-src/activity-stream-prerender.jsx]: I think string fallbacks work decently well at the moment (they fall back to English for any missing strings), although there are some cases where dynamic string ids get injected (e.g. TopStories) that could be handled better on a case-by-case basis

File: system-addon/content-src/activity-stream-prerender.jsx [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+      enabled: true,
+      icon: "pocket",
+      id: "topstories",
+      order: 1,
+      title: {id: "header_recommended_by", values: {provider: "Pocket"}},
+      read_more_endpoint: "https://getpocket.cdn.mozilla.net/explore/trending?src=ff_new_tab",
+      topics: [{name: "Must Reads", category: "World", url: "https://getpocket.com/explore/must-reads?src=ff_new_tab"}]
+    }
+  });
+
```
Comment (dmose) [inline system-addon/content-src/activity-stream-prerender.jsx]: Since each of the above dispatches effectively replaces ones that currently come from chrome, but have a comment above each one talking about how they differ or any assumptions in the interest of future code maintainability.

Comment (dmose) [reply system-addon/content-src/activity-stream-prerender.jsx]: Particularly keeping in mind that effectively this code may also need to be updated as the chrome code changes over time.

Comment (k88hudson) [reply system-addon/content-src/activity-stream-prerender.jsx]: It's not really so much that it comes from chrome as that specific actions are needed to set properties in certain reducers

Comment (k88hudson) [reply system-addon/content-src/activity-stream-prerender.jsx]: I can actually add tests to help with this

File: system-addon/content-src/activity-stream-prerender.jsx [hunk 15]
```diff
@@ -0,0 +1,63 @@
+const React = require("react");
```
Comment (dmose) [inline system-addon/content-src/activity-stream-prerender.jsx]: This file definitely wants unit tests.

File: system-addon/content-src/activity-stream.jsx [hunk 16]
```diff
@@ -6,11 +6,19 @@ const initStore = require("content-src/lib/init-store");
 const {reducers} = require("common/Reducers.jsm");
 const DetectUserSessionStart = require("content-src/lib/detect-user-session-start");
 const {addSnippetsSubscriber} = require("content-src/lib/snippets");
+const {actionTypes: at, actionCreators: ac} = require("common/Actions.jsm");
 
 new DetectUserSessionStart().sendEventOrAddListener();
 
-const store = initStore(reducers);
+const store = initStore(reducers, global.gActivityStreamPrerenderedState);
```
Comment (dmose) [inline system-addon/content-src/activity-stream.jsx]: The changes to this file want unit tests too.

File: system-addon/content-src/lib/init-store.js [hunk 17]
```diff
@@ -49,20 +49,23 @@ const messageMiddleware = store => next => action => {
  * @param  {object} reducers An object containing Redux reducers
```
Comment (dmose) [inline system-addon/content-src/lib/init-store.js]: There should probably be one or two unit tests added here.

File: system-addon/common/PrerenderData.jsm [hunk 18]
```diff
@@ -0,0 +1,42 @@
+const prefConfig = {
+  // Prefs listed with "invalidates: true" will prevent the prerendered version
+  // of AS from being used if their value is something other than what is listed
+  // here.
```
Comment (dmose) [inline system-addon/common/PrerenderData.jsm]: As we were discussing yesterday, some of these have minor consequences, which we currently believe are acceptable.  Documenting those consequences and tradeoffs here would help maintainability.

File: system-addon/content-src/activity-stream-prerender.jsx (hunk 14 above)
Comment (dmose) [reply system-addon/content-src/activity-stream-prerender.jsx]: That sounds good.

File: system-addon/test/unit/lib/PrefsFeed.test.js [hunk 19]
```diff
@@ -41,4 +45,38 @@ describe("PrefsFeed", () => {
...
+      feed.onAction({type: at.INIT});
+      assert.calledWith(feed._prefs.set, PRERENDER_PREF_NAME, false);
+    });
+  });
+  describe("onPrefChanged prerendering", () => {
+    it("should not change the prerender pref if the pref is not included in PrerenderData.invalidatingPrefs", () => {
+      feed.onPrefChanged("foo123", true);
+      assert.notCalled(feed._prefs.set);
+    });
+    it("should set the prerender pref if one of the prefs in PrerenderData.invalidatingPrefs is changed", () => {
```
Comment (dmose) [inline system-addon/test/unit/lib/PrefsFeed.test.js]: This wants to be split into two tests for the true -> false change and the false -> true change.

File: bin/render-activity-stream-html.js (hunk 12 above)
Comment (dmose) [reply bin/render-activity-stream-html.js]: Very nice; thanks!

File: bin/render-activity-stream-html.js [hunk 20]
```diff
@@ -1,3 +1,4 @@
+ /* eslint-disable no-sync */  /* eslint-disable no-sync */  /* eslint-disable no-sync */
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: Three times?  :-)

File: system-addon/content-src/components/Search/Search.jsx [hunk 21]
```diff
@@ -82,5 +82,5 @@ class Search extends React.Component {
 }
 
 // initialized is passed to props so that Search will rerender when it receives strings
```
Comment (dmose) [inline system-addon/content-src/components/Search/Search.jsx]: This comment needs to be updated too.

File: bin/render-activity-stream-html.js [hunk 22]
```diff
@@ -29,6 +29,7 @@ const DEFAULT_OPTIONS = {
  */
 function templateHTML(options, html) {
   const isPrerendered = !!html;
+  const scriptForPrerendering = `\n<script src="${options.baseUrl}data/content/activity-stream-initial-state.js"></script>`;
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: nit: It'd be even more readable if you just embedded a new line in the string rather than using /n.

File: bin/render-activity-stream-html.js [hunk 23]
```diff
@@ -44,8 +45,7 @@ function templateHTML(options, html) {
     <div id="root">${isPrerendered ? html : ""}</div>
     <div id="snippets-container">
       <div id="snippets"></div>
-    </div>${isPrerendered ? `
-    <script src="${options.baseUrl}data/content/activity-stream-initial-state.js"></script>` : ""}
+    </div>${isPrerendered ? `${scriptForPrerendering}` : ""}
```
Comment (dmose) [inline bin/render-activity-stream-html.js]: Much nicer; thanks!

File: bin/render-activity-stream-html.js (hunk 11 above)
Comment (dmose) [reply bin/render-activity-stream-html.js]: Great!

**Comment by k88hudson** [comment]
Ok, I filed a bunch of follow-ups and added them to this meta-bug: https://github.com/mozilla/activity-stream/issues/3372

File: system-addon/content-src/activity-stream-prerender.jsx (hunk 15 above)
Comment (k88hudson) [reply system-addon/content-src/activity-stream-prerender.jsx]: Done!

**Comment by k88hudson** [comment]
Filed a bugzilla ticket for the aboutNewTabService changes that need to land in mc:
https://bugzilla.mozilla.org/show_bug.cgi?id=1397875

File: system-addon/test/unit/activity-stream-prerender.test.jsx [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+const {PrerenderData} = require("common/PrerenderData.jsm");
+
+describe("prerenderStore", () => {
+  it("should create a store", () => {
+    const store = prerenderStore();
+    assert.isFunction(store.getState);
+  });
+  it("should start uninitialized", () => {
+    const store = prerenderStore();
+    const state = store.getState();
```
Comment (dmose) [inline system-addon/test/unit/activity-stream-prerender.test.jsx]: In all of these tests, calling getState wants to be part of the verification block, not the execution block.

File: system-addon/test/unit/activity-stream-prerender.test.jsx [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+
+    assert.equal(state.App.initialized, false);
+  });
+  it("should set the right locale and strings", () => {
+    const strings = {foo: "foo"};
```
Comment (dmose) [inline system-addon/test/unit/activity-stream-prerender.test.jsx]: The above line wants to be in its own setup block.

**Comment by dmose** [comment]
This looks great; r+, and feel free to fix any remaining nits without it going by me again.  Also, of the list of meta bugs, I assume that at least the rtl bug should be marked fx57 if things look visually wrong or jumpy on an RTL build.

File: system-addon/content-src/lib/init-store.js [hunk 26]
```diff
@@ -58,7 +58,17 @@ module.exports = function initStore(reducers, initialState) {
   );
 
   if (global.addMessageListener) {
+    let didInitialize = false;
```
Comment (k88hudson) [inline system-addon/content-src/lib/init-store.js]: I used this instead of `App.initialized` to decouple this from the specifics of the reducer's logic, in case it ever gets changed.

File: system-addon/content-src/lib/init-store.js [hunk 27]
```diff
@@ -47,22 +47,36 @@ const messageMiddleware = store => next => action => {
...
-    } catch (ex) {
-      console.error("Content msg:", msg, "Dispatch error: ", ex); // eslint-disable-line no-console
-      dump(`Content msg: ${JSON.stringify(msg)}\nDispatch error: ${ex}\n${ex.stack}`);
-    }
-  });
+  if (global.addMessageListener) {
+    let didInitialize = false;
+    global.addMessageListener(INCOMING_MESSAGE_NAME, msg => {
+      // We should discard any messages received before the initial rehydration
+      // message from Chrome, since that replaces the whole state anyway.
```
Comment (Mardak) [inline system-addon/content-src/lib/init-store.js]: nit: Firefox chrome is "chrome" ;)

Also, to get early lost `store.dispatch(ac.SendToMain({type: at.NEW_TAB_STATE_REQUEST})); `, on `INIT` here (i.e., activity stream just finished loading), re-dispatch STATE_REQUEST

File: system-addon/data/content/activity-stream.html [hunk 28]
```diff
@@ -1,24 +0,0 @@
...
-<html lang="en-us" dir="ltr">
-  <head>
-    <meta charset="utf-8">
-    <meta http-equiv="Content-Security-Policy-Report-Only" content="script-src 'unsafe-inline'; img-src http: https: data: blob:; style-src 'unsafe-inline'; child-src 'none'; object-src 'none'; report-uri https://tiles.services.mozilla.com/v4/links/activity-stream/csp">
-    <title></title>
-    <link rel="icon" type="image/png" id="favicon" href="chrome://branding/content/icon32.png"/>
-    <link rel="stylesheet" href="chrome://browser/content/contentSearchUI.css" />
-    <link rel="stylesheet" href="resource://activity-stream/data/content/activity-stream.css" />
-  </head>
-  <body class="activity-stream">
```
Comment (Mardak) [inline system-addon/data/content/activity-stream.html]: as_load_location is looking for `body.activity-stream`

**Comment by k88hudson** [comment]
@Mardak @dmose 

Ok, I added three more commits;

- 56df179 resolves the issue where a request can happen before activity stream has initialized; 
- c891f96 restores the class on the body element;
- 5614734 resolves the issue that @dmose and I talked that causes an empty locale to be rendered in the UI

**Comment by dmose** [comment]
Do we need/want to run this through Talos again, given that the performance characteristics have probably been changed a bit by the race fixes?

**Comment by Mardak** [comment]
This will need a patch to keep pine / m-c happy:
```diff
diff --git a/browser/base/content/test/static/browser_all_files_referenced.js b/browser/base/content/test/static/browser_all_files_referenced.js
--- a/browser/base/content/test/static/browser_all_files_referenced.js
+++ b/browser/base/content/test/static/browser_all_files_referenced.js
@@ -88,4 +88,8 @@ var whitelist = [
   {file: "resource://app/modules/NewTabWebChannel.jsm"},
 
+  // browser/extensions/activity-stream/data/content/activity-stream-prerendered.html
+  // will be used very soon by bug 1397875
+  {file: "resource://activity-stream/data/content/activity-stream-prerendered.html"},
+
   // layout/mathml/nsMathMLChar.cpp
   {file: "resource://gre/res/fonts/mathfontSTIXGeneral.properties"},
```

**Comment by Mardak** [comment]
Looks like this test is failing with this PR:
browser/components/search/test/browser_searchEngine_behaviors.js

Some brief debugging points to it waiting for and never getting content search events:
https://searchfox.org/mozilla-central/source/browser/components/search/test/browser_searchEngine_behaviors.js#96-107

**Comment by k88hudson** [comment]
Weird ok, I'll take a look

**Comment by k88hudson** [comment]
@Mardak @dmose ok, 5d8448 fixes the race condition where Search was getting loaded to early for the event to fire, as well as the unreferenced file.

File: mozilla-central-patches/fix-search-test-race.diff [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+   return ContentTask.spawn(browser, {}, async function(args) {
+-    return new Promise(resolve => {
+-      content.addEventListener("ContentSearchService", function listener(aEvent) {
+-        if (aEvent.detail.type == "State") {
+-          content.removeEventListener("ContentSearchService", listener);
+-          resolve();
+-        }
+-      });
+-    });
++    await ContentTaskUtils.waitForCondition(() => content.wrappedJSObject.gContentSearchController &&
```
Comment (dmose) [inline mozilla-central-patches/fix-search-test-race.diff]: It'd be more readable if each of the "content.wrappedJSObject" clauses had its own line, eslint-permitting.

Comment (dmose) [reply mozilla-central-patches/fix-search-test-race.diff]: The original style of test seems more robust, though I get that it wasn't quite right.  This fix seems ok for now, but I'd suggest also adding an XXX comment suggesting adding an event to the code or some such so that the test depends less on implementation details and is more useful for refactoring.

**Comment by dmose** [comment]
The aboutNewTabService patch to firefox.js seems to have bit-rotted.  Re-create it, using -U2 so that it will be more robust and likely to last a bit longer?

File: mozilla-central-patches/fix-search-test-race.diff (hunk 29 above)
Comment (k88hudson) [reply mozilla-central-patches/fix-search-test-race.diff]: I don't really see how this relies more on implementation details, given that that particular event has to be fired (and also in a particular order, i.e. after the listener is added in the test). Additionally, I kind of feel like using the event makes this test way harder to debug/step through 😜

Comment (dmose) [reply mozilla-central-patches/fix-search-test-race.diff]: OK, that's a reasonable enough rationale.

**Comment by Mardak** [comment]
There's a similar browser_google_behavior that has a "promiseContentSearchReady" but isn't failing ?
https://searchfox.org/mozilla-central/search?q=promiseContentSearchReady

**Comment by as-pine-proxy** [comment]
[Treeherder: 524b2f921428edc800ec2e33de4d85caf8719911](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=524b2f921428edc800ec2e33de4d85caf8719911)

**Comment by Mardak** [comment]
I reverted with https://github.com/mozilla/activity-stream/commit/ade80335b78346a740dd43922ea8f1edd00ba3f2 part of https://github.com/mozilla/activity-stream/commit/84534e4555deb97f6049c16ed496d3da5600f5e5 that added  mozilla-central-patches/fix-search-test-race.diff now that the export is on m-c with https://hg.mozilla.org/mozilla-central/rev/57f1221bf914
